presto-main/src/main/java/com/facebook/presto/operator/aggregation/arrayagg/SetUnionFunction.java

```diff
@@ -0,0 +1,162 @@
+
+    public SetUnionFunction()
+    {
+        super(NAME,
+                ImmutableList.of(typeVariable("T")),
+                ImmutableList.of(),
+                parseTypeSignature("array(T)"),
+                ImmutableList.of(parseTypeSignature("array(T)")));
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "return an array of unique values";
+    }
+
+    @Override
+    public InternalAggregationFunction specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionManager functionManager)
+    {
+        Type type = boundVariables.getTypeVariable("T");
+        ArrayType outputType = (ArrayType) typeManager.getParameterizedType(StandardTypes.ARRAY, ImmutableList.of(
+                TypeSignatureParameter.of(type.getTypeSignature())));
+        return generateAggregation(type, outputType);
+    }
+
+    private static InternalAggregationFunction generateAggregation(Type type, ArrayType outputType)
+    {
+        DynamicClassLoader classLoader = new DynamicClassLoader(SetUnionFunction.class.getClassLoader());
+
+        List<Type> inputTypes = ImmutableList.of(type);
+        AccumulatorStateSerializer<?> stateSerializer = new SetAggregationStateSerializer(outputType);
+        AccumulatorStateFactory<?> stateFactory = new SetAggregationStateFactory(type);
+
+        Type intermediateType = stateSerializer.getSerializedType();
+        List<ParameterMetadata> inputParameterMetadata = createInputParameterMetadata(type);
+        Class<? extends AccumulatorState> stateInterface = SetAggregationState.class;
+        AggregationMetadata metadata = new AggregationMetadata(
+                AggregationUtils.generateAggregationName(NAME, outputType.getTypeSignature(), inputTypes.stream().map(Type::getTypeSignature).collect(toImmutableList())),
+                inputParameterMetadata,
+                INPUT_FUNCTION.bindTo(type).bindTo(outputType),
+                COMBINE_FUNCTION,
+                OUTPUT_FUNCTION,
+                ImmutableList.of(new AccumulatorStateDescriptor(
+                        stateInterface,
+                        stateSerializer,
+                        stateFactory)),
+                outputType);
+
+        GenericAccumulatorFactoryBinder factory = AccumulatorCompiler.generateAccumulatorFactoryBinder(metadata, classLoader);
+        return new InternalAggregationFunction(NAME, inputTypes, ImmutableList.of(intermediateType), outputType, true, true, factory);
+    }
+
+    private static List<ParameterMetadata> createInputParameterMetadata(Type value)
+    {
+        return ImmutableList.of(new ParameterMetadata(STATE), new ParameterMetadata(NULLABLE_BLOCK_INPUT_CHANNEL, value), new ParameterMetadata(BLOCK_INDEX));
+    }
+
+    public static void input(Type type, Type arrayType, SetAggregationState state, Block inputBlock, int position)
```

Comment:
The type of `arrayType` should be `ArrayType`. Make it specific so the cast in line 127 is safe.

Reply:
Done